pkg/auth/token_handler.go

```diff
@@ -0,0 +1,259 @@
+	tokenRequest := authv1.TokenRequest{
+		Spec: authv1.TokenRequestSpec{
+			Audiences: []string{audience},
+		},
+	}
+
+	tokenRequestResponse, err := c.kubeClient.
+		CoreV1().
+		ServiceAccounts(serviceAccount.Namespace).
+		CreateToken(context.TODO(), serviceAccount.Name, &tokenRequest, metav1.CreateOptions{})
+
+	if err != nil {
+		return "", fmt.Errorf("could not request a token for %s: %w", serviceAccount, err)
+	}
+
+	return tokenRequestResponse.Status.Token, nil
+}
+
+// VerifyJWT verifies the given JWT for the expected audience and returns the parsed ID token.
+func (c *OIDCTokenHandler) VerifyJWT(ctx context.Context, jwt, audience string) (*IDToken, error) {
+	c.providerLock.RLock()
+	defer c.providerLock.RUnlock()
+
+	if c.provider == nil {
+		return nil, fmt.Errorf("provider is nil. Is the OIDC provider config correct?") // for devs this could also be a sign, that OIDCTokenHandler.Start() wasn't called
+	}
+
+	verifier := c.provider.Verifier(&oidc.Config{
+		ClientID: audience,
+	})
+
+	token, err := verifier.Verify(ctx, jwt)
+	if err != nil {
+		return nil, fmt.Errorf("could not verify JWT: %w", err)
+	}
+
+	return &IDToken{
+		Issuer:          token.Issuer,
+		Audience:        token.Audience,
+		Subject:         token.Subject,
+		Expiry:          token.Expiry,
+		IssuedAt:        token.IssuedAt,
+		AccessTokenHash: token.AccessTokenHash,
+	}, nil
+}
+
+func (c *OIDCTokenHandler) syncProviderConfigPeriodically(stopCh <-chan struct{}) {
+	ticker := time.NewTicker(c.providerSyncInterval)
+
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				if err := c.updateProviderConfig(); err != nil {
+					c.logger.Error("could not update OIDC provider config.", zap.Error(err))
+				}
+			case <-stopCh:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func (c *OIDCTokenHandler) updateProviderConfig() error {
+	featureFlags := c.featureStore.Load()
+	if !featureFlags.IsOIDCAuthentication() {
+		// we can skip if OIDC is not enabled
+		c.logger.Debug("oidc authentication is disabled, so skipping to update provider config")
+		return nil
+	}
+
+	c.logger.Debug("updating OIDC provider config...")
+
+	ctx := context.Background()
+
+	// load OIDC configuration from API server
+	oidcConfig, err := c.loadKubernetesOIDCConfiguration(ctx)
+	if err != nil {
+		return fmt.Errorf("could not get clusters OIDC configuration: %w", err)
+	}
+
+	if strings.HasPrefix(oidcConfig.Issuer, "https://kubernetes.default.svc") {
+		// setup http client for OIDC lib with the CA of the API server
+		httpClient, err := c.getHTTPClientForKubeAPIServer()
+		if err != nil {
+			return fmt.Errorf("could not get HTTP client with TLS certs of API server: %w", err)
+		}
+		ctx = oidc.ClientContext(context.Background(), httpClient)
+	} else {
+		// TODO: if OIDC issuer is an external one, configure http client with certs for external issuer
+		// (e.g. load certs from a config map)
+	}
+
+	// get OIDC provider
+	provider, err := oidc.NewProvider(ctx, oidcConfig.Issuer)
```

Comment:
There is a validation that the `given issuer != "the returned issuer"`  (the external issuer case), don't we need to handle that case?

https://github.com/coreos/go-oidc/blob/f0117a5f178cb36710dde42ede2154071b20ee94/oidc/oidc.go#L229-L235

Reply:
Added it